contents/docs/libraries/remix.md

```diff
@@ -18,36 +18,110 @@ yarn add posthog-js
 npm install --save posthog-js
 ```
 
-Then, go to `app/entry.client.tsx` and initialize PostHog as a component. You'll need both your API key and instance address (you can find these in your [project settings](https://us.posthog.com/project/settings)).
+First, you'll need to pass the `POSTHOG_API_KEY` to your client through Remix. We recommend placing them on `window.ENV` by following [this guide](https://remix.run/docs/en/main/guides/envvars#browser-environment-variables). You'll need both your API key and instance address (you can find these in your [project settings](https://us.posthog.com/project/settings)).
 
-```ts file=entry.client.tsx
-import { RemixBrowser } from "@remix-run/react";
-import { startTransition, StrictMode, useEffect } from "react";
-import { hydrateRoot } from "react-dom/client";
+Next, create a new PostHog context in `app/contexts/posthog-context.tsx`. This is necessary because of a [missing export statement](https://github.com/PostHog/posthog-js/issues/908) in `posthog-js`'s `package.json`. 
+
+```ts file=app/contexts/posthog-context.tsx
```

Comment:
This seems to cause a hydration error and de-opts the code into client side rendering for me. The follow code doesn't cause that problem. What do you think?

```
import posthog from "posthog-js";
import { useEffect, createContext, useContext, useRef } from "react";

type PosthogType = typeof posthog | undefined;

const PosthogContext = createContext<PosthogType>(undefined);

export function PosthogProvider({ children }: { children: React.ReactNode }) {
  const posthogInstanceRef = useRef<PosthogType>(undefined);

  useEffect(() => {

    // Only initialize PostHog once during component mount
    if (posthogInstanceRef.current) return;
    
    const apiKey = window.ENV?.POSTHOG_API_KEY;
    if (!apiKey) return;

    posthogInstanceRef.current = posthog.init(apiKey, {
      api_host: "https://us.i.posthog.com",
      loaded: (posthog) => {
        if (process.env.NODE_ENV === 'development') {
          posthog.debug();
        }
      },
    });
  }, []);

  return (
    <PosthogContext.Provider value={posthogInstanceRef.current}>
      {children}
    </PosthogContext.Provider>
  );
}

export const usePosthog = () => useContext(PosthogContext);
```

Reply:
Fair enough; React's recommendations around variables that must be initialized only once is pretty bad. 

One call out is that `process.env.NODE_ENV` is not available in the client, so I'll replace it with `window.ENV.IS_PROD` and add a note stating that this assumes the client ENV variables include `IS_PROD`. Note that also if the user follows Remix's recommendations on how to pass env variables to the client, `window.ENV` should always be defined, so we don't need the optional chaining.

Reply:
Also can you check again and/or push the remix repo you're testing with? I checked again, and I'm not getting the hydration error with the original code (that uses a function and a ref). I'm happy to change it to an effect and state, but given that React's recommendation is to use a function and ref, I'd prefer only changing it to effect and state only if it's not working. 

There's a bunch of different reasons why a hydration error is caused in Remix (including [3rd party browser extensions](https://www.jacobparis.com/content/remix-hydration-errors)); I just want to make sure React's recommendation doesn't work before making the change.

Reply:
It works fine on the first load, but subsequent loads get the hydration error. Tested on guest Chrome without extensions.

Here's the test repo: https://github.com/ivanagas/remix-cool2

Here's a test site where you can see the error on refresh: https://vercel.com/ivanagas-projects/remix-cool2

Reply:
Thanks for sharing the repo! Sorry about the delay; I was traveling. Downloaded, built, and ran it both in dev and as production. Unfortunately I haven't been able to replicate the hydration error in either my regular Chrome or incognito; or on the first load, uncached load, or reload: 

<img width="1604" alt="Screenshot 2024-12-10 at 5 02 43 PM" src="https://github.com/user-attachments/assets/ae112c56-65f2-4ae3-b5f6-4136dcd05ec5">

I'm not able to visit the vercel test site since I don't have access. Let me know if you're able to still reproduce the error, or if there are any other issues you notice.